common/components/common/projects/EditProjectForm.jsx

```diff
@@ -0,0 +1,188 @@
+import TagSelect from '../../common/tags/TagSelect.jsx'
+import DjangoCSRFToken from 'django-react-csrftoken'
+import ProjectAPIUtils from '../../../components/utils/ProjectAPIUtils.js';
+import type {ProjectDetailsAPIData} from '../../../components/utils/ProjectAPIUtils.js';
+import _ from 'lodash'
+
+type FormFields = {|
+  project_name: ?string,
+  project_location: ?string,
+  project_url: ?string,
+  project_description: ?string,
+  project_links: Array<LinkInfo>,
+  project_files: Array<FileInfo>,
+  project_thumbnail: FileInfo
+|};
+
+type Props = {|
+  projectId: number
+|};
+type State = {|
+  error: string,
+  formIsValid: boolean,
+  formFields: FormFields
+|};
+
+/**
+ * Encapsulates form for creating/editing projects
+ */
+class EditProjectForm extends React.PureComponent<Props,State> {
+  constructor(props: Props): void {
+    super(props);
+  
+    this.state = {
+      formIsValid: false,
+      formFields: {
+        project_name: "",
+        project_location: "",
+        project_url: "",
+        project_description: ""
+      }
+    };
+  }
+  
+  componentDidMount(): void {
+    if(this.props.projectId) {
+      ProjectAPIUtils.fetchProjectDetails(this.props.projectId, this.loadProjectDetails.bind(this), this.handleLoadProjectError.bind(this));
+    }
+  }
+  
+  loadProjectDetails(project: ProjectDetailsAPIData): void {
+    if(project.project_creator != window.DLAB_GLOBAL_CONTEXT.userID) {
+      this.setState({
+        error: "You are not authorized to edit this Project"
+      });
+    } else {
+      this.setState({
+        formFields: {
+          project_name: project.project_name,
+          project_location: project.project_location,
+          project_url: project.project_url,
+          project_description: project.project_description,
+          project_issue_area: project.project_issue_area && project.project_issue_area[0],
+          project_links: project.project_links,
+          project_files: project.project_files,
+          project_thumbnail: project.project_thumbnail
+        }
+      });
+      this.checkFormValidity();
+    }
+  }
+  
+  handleLoadProjectError(error: APIError): void {
+    this.setState({
+      error: "Failed to load project information"
+    });
+  }
+  
+  onFormFieldChange(formFieldName: string, event: SyntheticInputEvent<HTMLInputElement>): void {
+    this.state.formFields[formFieldName] = event.target.value;
+    this.forceUpdate();
+    this.checkFormValidity();
+  }
+  
+  onComponentChange(formFieldName: string, newValue: string): void {
+    this.state.formFields[formFieldName] = newValue;
+    this.checkFormValidity();
+  }
+  
+  checkFormValidity(): void {
+    var formFields = this.state.formFields;
+    
+    var valid = !_.isEmpty(formFields["project_name"]) && !_.isEmpty(formFields["project_description"]);
+    this.setState({
+      formIsValid: valid
+    });
+  }
+  
+  render(): React$Node {
+    if(this.state.error) {
```

Comment:
nit: can be terser with ternaries
`return this.state.error ? this._renderError() : this._renderForm();`

Reply:
Done